sql: ensure user has correct privileges when adding/removing regions

arulajmani opened the pull request:
Previously we did not account for privileges on database objects when
adding the default locality config on first region add or removing the
locality config on last region drop properly. In particular, we weren't
adding/removing the locality config on any descriptor that wasn't
visible to the user. This is bad because our validation logic expects
only and all objects in multi-region databases to have a valid locality
config. This means future accesses to such descriptors would fail
validation.

The root of this problem was the API choice here, `ForEachTableDesc`,
which filters out invisible descriptors. This patch instead switches
to using `forEachTableInMultiRegionDatabase`. While here, instead of
issuing separate requests for every table, I refactored this thing to
issue a single batch request instead.

Now that we view all the descriptors inside the database, unfiltered,
we perform privilege checks on them before proceeding with the add/drop
operation. In particular, the semantics are:
- admin users are allowed to add/drop regions as they wish.
- non admin-users require the CREATE privilege or must have ownership
on all the objects inside the database.

Closes #61003

Release note (sql change): `ALTER DATABASE .. SET PRIMARY REGION` now
requires both CREATE and ZONECONFIG privilege on all  objects inside
the database when adding the first region to the database. Same for
dropping the last region using `ALTER DATABASE ... DROP REGION`.

arulajmani committed:
sql: ensure user has correct privileges when adding/removing regions

Previously we did not account for privileges on database objects when
adding the default locality ocnfig on first region add or removing the
locality config on last region drop properly. In particular, we weren't
adding/removing the locality config on any descriptor that wasn't
visible to the user. This is bad because our validation logic expects
only and all objects in multi-region databases to have a valid locality
config. This means future accesses to such descriptors would fail
validation.

The root of this problem was the API choice here, `ForEachTableDesc`,
which filters out invisible descriptors. This patch instead switches
to using `forEachTableInMultiRegionDatabase`. While here, instead of
issuing separate requests for every table, I refactored this thing to
issue a single batch request instead.

Now that we view all the descriptors inside the database, unfiltered,
we perform privilege checks on them before proceeding with the add/drop
operation. In particular, the semantics are:
- admin users are allowed to add/drop regions as they wish.
- non admin-users require the CREATE privilege or must have ownership
on all the objects inside the database.

Closes #61003

Release note (sql change): `ALTER DATABASE .. SET PRIMARY REGION` now
requires both CREATE and ZONECONFIG privilege on all  objects inside
the database when adding the first region to the database. Same for
dropping the last region using `ALTER DATABASE ... DROP REGION`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62186)
<!-- Reviewable:end -->

otan reviewed (commented):
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @arulajmani)

___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region_privileges, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2EuN6-u2UcrFhluK-:-MW2EuN6-u2UcrFhluK0:b-rhuq1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/ccl/logictestccl/testdata/logic_test/multi_region_privileges#L25)):*
> ```
> 
> statement ok
> ALTER DATABASE  db SET PRIMARY REGION "us-east-1"
> ```

nit: extra space
___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region_privileges, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2F72T39PBx1a_iEsG:-MW2F72U3OxPzHkMvRJj:bxs1cg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/ccl/logictestccl/testdata/logic_test/multi_region_privileges#L35)):*
> ```
> 
> statement error user testuser does not have CREATE or ZONECONFIG privilege on t
> ALTER DATABASE db DROP REGION "us-east-1"
> ```

seems like we're missing a test on ADD REGION
___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region_privileges, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2F1NR3NgH4P9oOVJu:-MW2F1NR3NgH4P9oOVJv:b-rhuq1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/ccl/logictestccl/testdata/logic_test/multi_region_privileges#L57)):*
> ```
> 
> statement ok
> ALTER DATABASE  db SET PRIMARY REGION "us-east-1"
> ```

nit: extra space
___
*[pkg/sql/alter_database.go, line 371 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2FJJj78pZqQlwvLlv:-MW2FJJj78pZqQlwvLlw:bjdbeuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/sql/alter_database.go#L371)):*
> ```Go
> 	ctx context.Context, p *planner, tableDesc catalog.TableDescriptor,
> ) error {
> 	zoneConfigPrivErr := p.CheckPrivilege(ctx, tableDesc, privilege.ZONECONFIG)
> ```

i feel as if we should check each privilege and call out explicitly whether it is CREATE or ZONECONFIG that is the issue.
___
*[pkg/sql/region_util.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2FU6Y9uO_QEk8w30t:-MW2FU6Y9uO_QEk8w30u:bdzm2ip) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/sql/region_util.go#L66)):*
> ```Go
> 		p.txn,
> 		sessiondata.InternalExecutorOverride{
> 			User: security.RootUserName(),
> ```

can you add a comment about why root is needed.
___
*[pkg/sql/type_change.go, line 487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2GMcNCTGxvzSOdVSZ:-MW2GMcO0KknAtDeEsXI:b-kqhllo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/sql/type_change.go#L487)):*
> ```Go
> 	)
> 	defer cleanup()
> 	localPlanner := p.(*planner)
> ```

do we need a privilege check here as well, given that RBR tables are having zone configs attached?
looks like the checks are only in for final DROP or initial ADD


<!-- Sent from Reviewable.io -->

otan commented:
chatted online and ADD/DROP non-first region not included, but curious about this test failure

```

        (XXUUU) PARTITION us-east-1: enum value "us-east-1" is not yet public
        datum.go:4026: in MakeDEnumFromLogicalRepresentation()

``

RichardJCai commented on pkg/sql/alter_database.go:
```diff
@@ -358,6 +360,24 @@ func (p *planner) AlterDatabaseDropRegion(
+) error {
+	zoneConfigPrivErr := p.CheckPrivilege(ctx, tableDesc, privilege.ZONECONFIG)
+	createPrivErr := p.CheckPrivilege(ctx, tableDesc, privilege.CREATE)
+	if zoneConfigPrivErr != nil && createPrivErr != nil {
```
Since we eventually want to be more accurate with Postgres privileges and they don't have `CREATE` privilege for tables, maybe we should only show in the error that the user doesn't have `ZONECONFIG`.  Or maybe we should show the same error message for ALTER TABLE LOCALITY `must be owner of table %s or have CREATE privilege on table %s`

Actually would it make more sense to pull out the privilege check in `alter_table_locality.go` (and possibly even `alter_table.go` into a helper function and just re-use it here so we don't duplicate privilege checks that should technically be the same?

arulajmani replied:
> Since we eventually want to be more accurate with Postgres privileges and they don't have CREATE privilege for tables, maybe we should only show in the error that the user doesn't have ZONECONFIG. 

If we're accepting CREATE as a valid privilege to do this operation, then IMO the error message should indicate that. If we want to move more in line with Postgres then maybe we shouldn't accept CREATE here at all to begin with. Would you prefer that?

> Or maybe we should show the same error message for ALTER TABLE LOCALITY must be owner of table %s or have CREATE privilege on table %s

I think that thing is busted to be honest, it should accept ZONECONFIG as a valid privilege too. I was thinking of calling this function in `alter_table_locality.go`,  there's a TODO, I'll just address it here. That being said, I think it makes sense to add a nod to the "owner" part in the error message. 

> Actually would it make more sense to pull out the privilege check in alter_table_locality.go (and possibly even alter_table.go into a helper function and just re-use it here so we don't duplicate privilege checks that should technically be the same?

But aren't the privilege checks different for `alter_table_locality.go` and `alter_table.go`? IMO `alter_table_locality.go` should accept ZONECONFIG but `alter_table.go` should just be CREATE.

RichardJCai replied:
I was thinking we could pull out some of the shared logic for the error check from `alter_table.go` and `alter_table_locality.go` and use it in all three places. Something like one helper function for checking `CREATE` or `OWNER` which is used in all 3, however for `alter_table_locality.go` and this function here, we can wrap that helper and add an additional check for `ZONECONFIG`. If this turns out to be messy then I'd be okay with just creating one function for the locality checks.

arulajmani replied:
> there's a TODO, I'll just address it here. That being said, I think it makes sense to add a nod to the "owner" part in the error message.

Changed my mind, I'll address this in a separate commit. It needs some testing as well.

arulajmani replied:
Ah, I see what you mean, I'm on board. I'll do it when I address the `alter_table_locality.go` stuff :+1:

RichardJCai replied:
Do we have an issue for the `alter_table_locality.go` privilege stuff or a general issue on the MR privilege stuff?

arulajmani replied:
I did the alter_table_locality in the second commit. As for other MR privileges stuff, I still have to write them out after the whole Slack discussion.

arulajmani replied:
Followups for after this thing goes in: 

https://github.com/cockroachdb/cockroach/issues/62252
https://github.com/cockroachdb/cockroach/issues/62250

arulajmani commented:
> chatted online and ADD/DROP non-first region not included so most comments can be ignored, but curious about this test failure:
       (XXUUU) PARTITION us-east-1: enum value "us-east-1" is not yet public
        datum.go:4026: in MakeDEnumFromLogicalRepresentation()


Looks like when we GetAllDescriptors, we hydrate tables using the type descriptors read from the store -- not the uncommitted type descriptors modified previously in the transaction. Put another way, we're not seeing the promotion changes we just made. The naive fix is trivial, but I want to think about this a bit more.

arulajmani commented:
> Looks like when we GetAllDescriptors, we hydrate tables using the type descriptors read from the store -- not the uncommitted type descriptors modified previously in the transaction. Put another way, we're not seeing the promotion changes we just made. The naive fix is trivial, but I want to think about this a bit more.

Actually, turns out we weren't running the batch that modified the type descriptors until after repartitioning. It didn't matter earlier, but it does now because of how `GetAllDescriptors` hydrates things.

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @otan, and @RichardJCai)

___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region_privileges, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2F72T39PBx1a_iEsG:-MW5E71n92d7FDumAxPX:bhrlhqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/ccl/logictestccl/testdata/logic_test/multi_region_privileges#L35)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

seems like we're missing a test on ADD REGION
</blockquote></details>

Upcoming patch. 
___
*[pkg/sql/alter_database.go, line 371 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2FJJj78pZqQlwvLlv:-MW5CsAn978hzRi3r7ih:bpjnp83) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/sql/alter_database.go#L371)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i feel as if we should check each privilege and call out explicitly whether it is CREATE or ZONECONFIG that is the issue.
</blockquote></details>

I don't fully follow -- you need either of those privileges, not both.
___
*[pkg/sql/region_util.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2FU6Y9uO_QEk8w30t:-MW5E1U3BJd4szd5cIdr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/sql/region_util.go#L66)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can you add a comment about why root is needed.
</blockquote></details>

Done.
___
*[pkg/sql/type_change.go, line 487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2GMcNCTGxvzSOdVSZ:-MW5D7WT2s0I9ZprJYKx:b-hhef6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/sql/type_change.go#L487)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

do we need a privilege check here as well, given that RBR tables are having zone configs attached?
looks like the checks are only in for final DROP or initial ADD
</blockquote></details>

Upcoming patch!  


<!-- Sent from Reviewable.io -->

arulajmani committed:
sql: refactor alter table locality privileges check

This patch calls into the `ensureCorrectMultiRegionPrivilegesForTable`
function instead of checking privileges. This doesn't change any
behavior, but I've added tests ensuring the semantics are sane.

Release note: None

arulajmani reviewed (commented):
There's a couple fundamental changes since I put this out:
- Instead of accepting ZONECONFIG or CREATE, now we only accept CREATE. This is from the discussion thread on Slack. 
- There's a second commit that basically adds testing for ALTER TABLE LOCALITY. Functionally, nothing has changed here. 

RFAL. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-:-MW6pAjw8L3Y0_LhdJm9:b-ympa09)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, @otan, and @RichardJCai)



<!-- Sent from Reviewable.io -->

RichardJCai reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, @otan, and @RichardJCai)

___
*[pkg/sql/alter_database.go, line 370 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MWALtnd-qt1DwDtpCIp:-MWALtnd-qt1DwDtpCIq:bry9o14) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ea0538994aff81db39168b34582223075f97e79/pkg/sql/alter_database.go#L370)):*
> ```Go
> // - or have the CREATE privilege on the table.
> // privilege on the table descriptor.
> func (p *planner) ensureCorrectMultiRegionPrivilegesForTable(
> ```

Should this be renamed to checkCanAlterLocality or something along those lines to make it more specific.

Also I don't know if this is really a rule but I think we generally call these privilege check functions "check..."
___
*[pkg/sql/alter_database.go, line 414 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MWAMcL1B59nuWSnA1bt:-MWAMcL1B59nuWSnA1bu:b-fyl1pe) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ea0538994aff81db39168b34582223075f97e79/pkg/sql/alter_database.go#L414)):*
> ```Go
> 
> 		switch t := tbDesc.LocalityConfig.Locality.(type) {
> 		case *descpb.TableDescriptor_LocalityConfig_Global_:
> ```

I don't understand enough on this switch statement to review here, maybe @otan or @ajstorm can take a look at this specific part?
___
*[pkg/sql/alter_database.go, line 378 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MWAOjOK1bYa7oQwYtDu:-MWAOjOK1bYa7oQwYtDv:bhmjyo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a508f4f0a2fce579a6f73d695decc7cbb91fe92/pkg/sql/alter_database.go#L378)):*
> ```Go
> 	}
> 	if !hasAdminRole {
> 		err := p.CheckPrivilege(ctx, tableDesc, privilege.CREATE)
> ```

Can you add a TODO here that the `CREATE` priv check is not inline with PG and should be removed when we refactor the privilege system
___
*[pkg/sql/type_change.go, line 487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2GMcNCTGxvzSOdVSZ:-MWANjnUCfhTNYkfAhVc:b-ox9gt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/sql/type_change.go#L487)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Upcoming patch!  
</blockquote></details>

Just double checking, this happens in the `exec` function so the correct privilege checks are done by this point right?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @otan, and @RichardJCai)

___
*[pkg/sql/alter_database.go, line 370 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MWALtnd-qt1DwDtpCIp:-MWA_wUvA3tLQo8_1OpH:b9fg94o) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ea0538994aff81db39168b34582223075f97e79/pkg/sql/alter_database.go#L370)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Should this be renamed to checkCanAlterLocality or something along those lines to make it more specific.

Also I don't know if this is really a rule but I think we generally call these privilege check functions "check..."
</blockquote></details>

`checkCanAlterLocality` isn't really true though, as the same rules apply when re-partitioning a multi-region table because of a region add (more on this below). I'm tentatively changing this to `checkPrivilegesForMultiRegionOperation`, lemme know what you think. 
___
*[pkg/sql/alter_database.go, line 414 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MWAMcL1B59nuWSnA1bt:-MWAbBc3EiIp3gIrhTNP:b-y3s4ih) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ea0538994aff81db39168b34582223075f97e79/pkg/sql/alter_database.go#L414)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I don't understand enough on this switch statement to review here, maybe @otan or @ajstorm can take a look at this specific part?
</blockquote></details>

Rebase conflict, there's no functional change here. 
___
*[pkg/sql/alter_database.go, line 378 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MWAOjOK1bYa7oQwYtDu:-MWAbAy5-H-lpU20k0p5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a508f4f0a2fce579a6f73d695decc7cbb91fe92/pkg/sql/alter_database.go#L378)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Can you add a TODO here that the `CREATE` priv check is not inline with PG and should be removed when we refactor the privilege system
</blockquote></details>

Done.
___
*[pkg/sql/type_change.go, line 487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MW2GMcNCTGxvzSOdVSZ:-MWAbSRv78Qbz5SvwhqY:bm6r03t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1cb9bf82069c607770980ba5038b5bf771db2/pkg/sql/type_change.go#L487)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Just double checking, this happens in the `exec` function so the correct privilege checks are done by this point right?
</blockquote></details>

Not exactly -- right now, only the first region add/last region drops perform a privilege check (in the `exec` function). The check happens on all tables in the database as these are special operations which involve modifying the locality config on all objects inside the database. 

All subsequent region adds/drops don't affect all the objects inside the database -- they only affect these "regional by row" tables. Regional by row tables need to be repartitioned based on the new set of regions and a zone config is then applied for every partition. I field https://github.com/cockroachdb/cockroach/issues/62250 to capture the work required for this scenario.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (approved):
Privileges LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-:-MWQ43aGCbRrau1q1k91:bxe9f59)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @otan, and @RichardJCai)

___
*[pkg/sql/alter_database.go, line 414 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62186#-MWAMcL1B59nuWSnA1bt:-MWQ3d2b5AOA8HNKpciS:btn5jp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ea0538994aff81db39168b34582223075f97e79/pkg/sql/alter_database.go#L414)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Rebase conflict, there's no functional change here. 
</blockquote></details>

Gotcha


<!-- Sent from Reviewable.io -->

arulajmani commented:
TFTR! In @otan's words, if it's good enough for Richard its good enough for me. 

bors r=richardjcai
